Userland/Services/WebDriver/Session.cpp

```diff
@@ -899,6 +900,78 @@ ErrorOr<JsonValue, WebDriverError> Session::get_element_tag_name(JsonValue const
+    // 4. If args is not an Array return error with error code invalid argument.
+    if (!properties.has_array("args"sv))
+        return WebDriverError::from_code(ErrorCode::InvalidArgument, "Payload doesn't have an 'args' string property");
+    auto const& args = properties.get("args"sv).as_array();
+
+    // 5. Let arguments be the result of calling the JSON deserialize algorithm with arguments args.
+    // NOTE: We forward the JSON array to the Browser and then WebContent process over IPC, so this is not necessary.
+
+    // 6. Return success with data script and arguments.
+    return ScriptArguments { script, args };
+}
+
+// 13.2.1 Execute Script, https://w3c.github.io/webdriver/#dfn-execute-script
+ErrorOr<JsonValue, WebDriverError> Session::execute_script(JsonValue const& payload)
+{
+    // 1. Let body and arguments be the result of trying to extract the script arguments from a request with argument parameters.
+    auto const& [body, arguments] = TRY(extract_the_script_arguments_from_a_request(payload));
+
+    // 2. If the current browsing context is no longer open, return error with error code no such window.
+    TRY(check_for_open_top_level_browsing_context_or_return_error());
+
+    // FIXME: 3. Handle any user prompts, and return its value if it is an error.
+
+    // 4., 5.1-5.3.
+    Vector<String> json_arguments;
+    arguments.for_each([&](JsonValue const& json_value) {
+        // NOTE: serialized() instead of to_string() ensures proper quoting.
+        json_arguments.append(json_value.serialized<StringBuilder>());
+    });
+
+    dbgln("Executing script with 'args': [{}] / 'body':\n{}", String::join(", "sv, json_arguments), body);
+    auto result = m_browser_connection->execute_script(body, json_arguments, m_timeouts_configuration.script_timeout, false);
+    dbgln("Executing script returned: {}", result.json_result());
+
+    switch (result.result_type()) {
+    // 6. If promise is still pending and the session script timeout is reached, return error with error code script timeout.
+    case Web::WebDriver::ExecuteScriptResultType::Timeout:
+        return WebDriverError::from_code(ErrorCode::ScriptTimeoutError, "Script timed out");
+    // 7. Upon fulfillment of promise with value v, let result be a JSON clone of v, and return success with data result.
+    case Web::WebDriver::ExecuteScriptResultType::PromiseResolved:
+        // NOTE: This is assumed to be a valid JSON value.
+        return MUST(JsonValue::from_string(result.json_result()));
+    // 8. Upon rejection of promise with reason r, let result be a JSON clone of r, and return error with error code javascript error and data result.
+    // FIXME: The definition of 'error' doesn't mention how to return data!
+    case Web::WebDriver::ExecuteScriptResultType::PromiseRejected:
+    case Web::WebDriver::ExecuteScriptResultType::JavaScriptError:
+        return WebDriverError::from_code(ErrorCode::JavascriptError, "Script returned an error");
```

Comment:
Does `json_result` have anything useful here if we get an error? If so, we can:
```c++
return WebDriverError::from_code(ErrorCode::JavascriptError, String::formatted("Script returned an error: {}", json_result));
```

Reply:
Resolved the FIXME instead to do it properly (as a `"data"` key of the returned error JSON object)